server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,310 @@
+        boolean success = false;
+        try {
+            final SnapshotShard snapshotShard = queue.poll(0L, TimeUnit.MILLISECONDS);
+            if (snapshotShard != null) {
+                threadPool.executor(executorName()).execute(new AbstractRunnable() {
+                    @Override
+                    protected void doRun() {
+                        if (clusterService.state().nodes().isLocalNodeElectedMaster() == false) {
+                            logger.debug("skipping snapshot shard size retrieval for {} as node is no longer master", snapshotShard);
+                            return;
+                        }
+
+                        final RepositoriesService repositories = repositoriesService.get();
+                        assert repositories != null;
+                        final Repository repository = repositories.repository(snapshotShard.snapshot.getRepository());
+
+                        logger.debug("fetching snapshot shard size for {}", snapshotShard);
+                        final IndexShardSnapshotStatus status = repository.getShardSnapshotStatus(
+                            snapshotShard.snapshot().getSnapshotId(),
+                            snapshotShard.index(),
+                            snapshotShard.shardId()
+                        );
+
+                        final long snapshotShardSize = status.asCopy().getTotalSize();
+                        logger.debug("snapshot shard size for {}: {} bytes", snapshotShard, snapshotShardSize);
+
+                        boolean updated = false;
+                        synchronized (InternalSnapshotsInfoService.this) {
+                            if (isMaster) {
+                                final ImmutableOpenMap.Builder<SnapshotShard, Long> newSnapshotShardSizes =
+                                    ImmutableOpenMap.builder(knownSnapshotShardSizes);
+                                updated = newSnapshotShardSizes.put(snapshotShard, snapshotShardSize) == null;
+                                knownSnapshotShardSizes = newSnapshotShardSizes.build();
+                            }
+                        }
+                        if (updated) {
+                            rerouteService.get().reroute("snapshot shard size updated", Priority.HIGH,
+                                ActionListener.wrap(
+                                    r -> logger.trace("reroute after snapshot shard size update completed"),
+                                    e -> logger.debug("reroute after snapshot shard size update failed", e)));
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.warn(() -> new ParameterizedMessage("failed to retrieve shard size for {}", snapshotShard), e);
+                    }
+
+                    @Override
+                    public void onAfter() {
+                        unknownSnapshotShards.remove(snapshotShard);
```

Comment:
> but the thread is halted before removing from 

This is impossible for a running node isn't it? We don't ever kill/interrupt actively executing threads on the generic pool and then keep going with the node? 
(can you even halt a thread in such a way that this method which is called in a `finally` block isn't executed every time we run this `Runnable`?)

Reply:
I'm interested in Armin's questions answers, but besides this Henning raised a good point where the snapshot shard size should be removed from `unknownSnapshotShards` as soon as it is read, before it is added back to `knownSnapshotShards`

Reply:
I should probably have used "paused" rather than "halted". The finally block does run, but steps 3 and 4 goes in before it runs. Thanks for addressing this.